src/index.ts

```diff
@@ -74,253 +74,272 @@ const TYPED_MESSAGE_SCHEMA = {
+ * @param {Object} types - Type definitions
+ * @returns {Buffer} - Hash of an object
+ */
+function hashType(
+  primaryType: string,
+  types: Record<string, MessageTypeProperty[]>,
+): Buffer {
+  return ethUtil.keccak(encodeType(primaryType, types));
+}
 
-  /**
-   * Removes properties from a message object that are not defined per EIP-712
-   *
-   * @param {Object} data - typed message object
-   * @returns {Object} - typed message object with only allowed fields
-   */
-  sanitizeData<T extends MessageTypes>(
-    data: TypedData | TypedMessage<T>,
-  ): TypedMessage<T> {
-    const sanitizedData: Partial<TypedMessage<T>> = {};
-    for (const key in TYPED_MESSAGE_SCHEMA.properties) {
-      if (data[key]) {
-        sanitizedData[key] = data[key];
-      }
-    }
-    if ('types' in sanitizedData) {
-      sanitizedData.types = { EIP712Domain: [], ...sanitizedData.types };
+/**
+ * Removes properties from a message object that are not defined per EIP-712
+ *
+ * @param {Object} data - typed message object
+ * @returns {Object} - typed message object with only allowed fields
+ */
+function sanitizeData<T extends MessageTypes>(
+  data: TypedData | TypedMessage<T>,
+): TypedMessage<T> {
+  const sanitizedData: Partial<TypedMessage<T>> = {};
+  for (const key in TYPED_MESSAGE_SCHEMA.properties) {
+    if (data[key]) {
+      sanitizedData[key] = data[key];
     }
-    return sanitizedData as Required<TypedMessage<T>>;
-  },
+  }
+  if ('types' in sanitizedData) {
+    sanitizedData.types = { EIP712Domain: [], ...sanitizedData.types };
+  }
+  return sanitizedData as Required<TypedMessage<T>>;
+}
 
-  /**
-   * Signs a typed message as per EIP-712 and returns its keccak hash
-   *
-   * @param {Object} typedData - Types message data to hash as per eip-712
-   * @returns {Buffer} - keccak hash of the resulting signed message
-   */
-  eip712Hash<T extends MessageTypes>(
-    typedData: Partial<TypedData | TypedMessage<T>>,
-    useV4 = true,
-  ): Buffer {
-    const sanitizedData = this.sanitizeData(typedData);
-    const parts = [Buffer.from('1901', 'hex')];
+/**
+ * Signs a typed message as per EIP-712 and returns its keccak hash
+ *
+ * @param {Object} typedData - Types message data to hash as per eip-712
+ * @returns {Buffer} - keccak hash of the resulting signed message
+ */
+function eip712Hash<T extends MessageTypes>(
+  typedData: TypedData | TypedMessage<T>,
+  useV4 = true,
+): Buffer {
+  const sanitizedData = sanitizeData(typedData);
+  const parts = [Buffer.from('1901', 'hex')];
+  parts.push(
+    hashStruct(
+      'EIP712Domain',
+      sanitizedData.domain,
+      sanitizedData.types,
+      useV4,
+    ),
+  );
+  if (sanitizedData.primaryType !== 'EIP712Domain') {
     parts.push(
-      this.hashStruct(
-        'EIP712Domain',
-        sanitizedData.domain,
+      hashStruct(
+        // TODO: Validate that this is a string, so this type cast can be removed.
+        sanitizedData.primaryType as string,
```

Comment:
Here's the second type assertion that was required. At first it looked like this was already guaranteed to be a string, because `primaryType` is a keyof the `MessageTypes` object, and that's string-indexed. But apparently TypeScript is screwy when it comes to index types, and it happily allows using a `number` index as well (I tested it to be sure), despite this attempt at preventing that. So technically this could be a number. But that's clearly against the spec, and it's clearly what we were trying to prevent with that type.

We can shore this up later with runtime validation. Or better types. Or both.